lambda/translcore.ml

```diff
@@ -322,36 +329,155 @@ let can_apply_primitive p pmode pos args =
+      | [] -> (kind, transl None e)
+      | _ ->
+        let result_var = Ident.create_local "after_voids" in
+        (kind,
+         Llet (Strict, kind, result_var,
+               transl None e,
+               List.fold_left (fun lam e ->
+                 catch_void (fun void_k -> transl void_k e) lam kind)
+                 (Lvar result_var) after))
+    in
+    let transl_with_voids_before kind lam before =
+      List.fold_left (fun lam e -> catch_void (fun void_k -> transl void_k e)
+                                     lam kind)
+        lam before
+    in
+    let (kind, lam) = transl_with_voids_after e after in
+    (transl_with_voids_before kind lam before, kind)
+  in
+  let group_voids exprs =
+    (* We need to group the voids with an element before them or the element
+       after them.  We prefer to do them as part of an element after them in
+       eval order, because doing them "before" an element is free while the
+       alternative adds a let binding. *)
+    let rec get_afters idx afters = function
+      | e :: exprs when is_void e ->
+        get_afters (idx + 1) (e :: afters) exprs
+      | exprs -> (idx,exprs,afters)
+    in
+    let rec group idx befores nonvoid afters acc exprs =
+      match exprs with
+      | [] -> (List.rev befores,nonvoid, List.rev afters) :: acc
+      | e :: exprs when is_void e ->
+        group (idx+1) (e :: befores) nonvoid afters acc exprs
+      | e :: exprs ->
+        group (idx+1) [] e []
+          ((List.rev befores,nonvoid,List.rev afters) :: acc) exprs
+    in
+    let (idx,exprs,afters) = get_afters 0 [] exprs in
+    let (first_non_void,exprs) = match exprs with
+      | [] -> assert false
+      | (e :: exprs) -> e,exprs
+    in
+    List.rev (group (idx+1) [] first_non_void afters [] exprs)
+  in
+  let grouped = group_voids expr_list in
+  List.split (List.map transl_with_voids_before_and_after grouped)
+
+let rec transl_exp ~scopes void_k e =
+  transl_exp1 ~scopes ~in_new_scope:false void_k e
 
 (* ~in_new_scope tracks whether we just opened a new scope.
 
    We go to some trouble to avoid introducing many new anonymous function
    scopes, as `let f a b = ...` is desugared to several Pexp_fun.
 *)
-and transl_exp1 ~scopes ~in_new_scope e =
+and transl_exp1 ~scopes ~in_new_scope void_k e =
   List.iter (Translattribute.check_attribute e) e.exp_attributes;
   let eval_once =
     (* Whether classes for immediate objects must be cached *)
     match e.exp_desc with
       Texp_function _ | Texp_for _ | Texp_while _ -> false
     | _ -> true
   in
-  if eval_once then transl_exp0 ~scopes ~in_new_scope  e else
-  Translobj.oo_wrap e.exp_env true (transl_exp0 ~scopes ~in_new_scope) e
+  if eval_once then transl_exp0 ~scopes ~in_new_scope void_k e else
+  Translobj.oo_wrap e.exp_env true (transl_exp0 ~scopes ~in_new_scope void_k) e
 
-and transl_exp0 ~in_new_scope ~scopes e =
+(* [void_k] is [Some n] iff [e] has layout void, where [n] is a static handler
```

Comment:
I agree this was confusing, but I confess that the idea of copy-pasting a comment 15 times did not spark joy.  So instead I've tried to improve the API to make the code itself clearer:

I added a new type:

```ocaml
type void_continuation = Void_cont of int | Not_void
```

I replaced all the int options with this type, and moved the type to `lambda.mli` so it can be shared in the `transl*` modules.  I also moved the comment explaining the system, previously on `Translcore.value_kind_if_not_void`, to the new type in `lambda.mli`.